.drone.yml

```diff
@@ -3085,9 +3129,14 @@ services:
 volumes:
   - name: dockersock
     temp: {}
+  # this persistent volume caches RPMs near Drone so that we don't need to download the
+  # entire repo contents from S3 every time to build the repo, we just sync any differences
+  - name: rpmrepo
+    claim:
+      name: drone-s3-rpmrepo-pvc
```

Comment:
I'm not exactly the most familiar with the PVC/PV stuff in kubernetes as we don't get much use in the on prem clusters. I'm not sure what backing store is planned to be used, but the cloud ones do tend to create some considerations.

AFAIK the GCE/AWS pv backends work as EBS volumes, which means the volume once created will only live in a single AZ. This potentially makes the pod intolerant to AZ failure, if that's important. It likely can only be bound RW once as well, which would presumably synchronize if trying to publish multiple releases (I'm not sure if drone does something special here). This may not be important, IDK.

While the proposed change is probably the more correct way, specific to caching, an alternate approach I would consider would be to mount the local ephemeral disks from the instance as a cache. Whenever the instance is recycled it would be cold, and each node would have it's own cache which might increase miss rates. The downside I would foresee is I don't know if building the RPM repo is safe if another sync or publish is also modifying the filesystem at the same time, so might still need some coordination between parallel executions.

Reply:
Yeah, you're correct here. The backing store is a `gcePersistentDisk` which can only be mounted `ReadWrite` by one pod at a time. This RPM repo publishing job only runs once for every Teleport version tag that we promote via Drone, so it should be reasonably concurrency-safe.

The one time I could see there being an issue would be when we're publishing multiple security backports in a short space of time, but the effect _should_ be that if multiple pods hit this stage of the pipeline at once:
- one pod will win and mount the volume, syncing its particular tag version to the cache and then uploading to S3
- the other pods will fail to launch (because they can't multi-attach the volume) and will retry with backoff
- when the volume is released, one pod will win again and mount the volume, syncing its own tag version to the cache (which will be different to previous versions) and completing an upload
- repeat this process until all pods have run successfully

As the RPM publishing step takes around 30 seconds total, this should be well within all timeouts for Drone in general.

The AZ issue is something I hadn't really thought of, but Drone isn't running in HA currently so this isn't something I've seen a big issue with. It's definitely a future concern.

Ephemeral storage is another potential solution, although we do see that an extra GKE node is often launched via autoscaling when we run publishing jobs because of the high number of concurrent pipelines running, so the chances are that this cache would often be cold and result in GBs of RPMs needing to be synced from S3. We can keep an eye on the usage of this job over time and see whether it'd be more appropriate to use ephemeral storage.